ui/core/components/suggest_gems_action.ts

```diff
@@ -0,0 +1,627 @@
+			return updatedGear;
+		}
+
+		// If we exceeded the stat cap, then work backwards through the socket list and replace each gem with the next highest priority option until we are below the cap
+		const nextGem = this.sim.db.lookupGem(gemCaps[numPasses + 1].gemId);
+		const nextCap = gemCaps[numPasses + 1].statCaps;
+		let capForReplacement = currentCap.subtract(nextCap);
+
+		if (currentCap.computeEP(capForReplacement) <= 0) {
+			capForReplacement = currentCap;
+		}
+
+		for (var idx = socketList.length - 1; idx >= firstIdx; idx--) {
+			if (newStats.belowCaps(capForReplacement)) {
+				break;
+			}
+
+			updatedGear = updatedGear.withGem(socketList[idx].itemSlot, socketList[idx].socketIdx, nextGem);
+			newStats = await this.updateGear(updatedGear);
+		}
+
+		// Now run a new pass to check whether we've exceeded the next stat cap
+		let nextIdx = idx + 1;
+
+		if (!newStats.belowCaps(currentCap)) {
+			nextIdx = firstIdx;
+		}
+
+		return await this.fillGemsToCaps(updatedGear, socketList, gemCaps, numPasses + 1, nextIdx);
+	}
+
+	abstract activateMetaGem(gear: Gear): Gear;
+
+	abstract updateGemPriority(ungemmedGear: Gear, passiveStats: Stats): void;
+
+	abstract allowGemInSocket(gemColor: GemColor, socketColor: GemColor, itemSlot: ItemSlot, item: EquippedItem): boolean;
+}
+
+export class PhysicalDPSGemOptimizer extends GemOptimizer {
+	metaGemID: number = 41398; // Relentless Earthsiege Diamond
+	arpSlop: number = 11;
+	expSlop: number = 4;
+	hitSlop: number = 4;
+	useArpGems: boolean;
+	useExpGems: boolean;
+	useAgiGems: boolean;
+	useStrGems: boolean;
+	arpTarget!: number;
+	passiveArp!: number;
+	arpStackDetected!: boolean;
+	tearSlot!: ItemSlot | null;
+
+	constructor(simUI: IndividualSimUI<any>, useArpGems: boolean, useExpGems: boolean, useAgiGems: boolean, useStrGems: boolean) {
+		super(simUI);
+		this.useArpGems = useArpGems;
+		this.useExpGems = useExpGems;
+		this.useAgiGems = useAgiGems;
+		this.useStrGems = useStrGems;
+	}
+
+	updateGemPriority(ungemmedGear: Gear, passiveStats: Stats) {
+		// First calculate any gear-dependent stat caps.
+		this.arpTarget = this.calcArpTarget(ungemmedGear);
+		const arpCap = new Stats().withStat(Stat.StatArmorPenetration, this.arpTarget + this.arpSlop);
+		const critCap = this.calcCritCap(ungemmedGear);
+		const expCap = new Stats().withStat(Stat.StatExpertise, this.calcExpTarget() + this.expSlop);
+		const hitCap = new Stats().withStat(Stat.StatMeleeHit, 8. * 32.79 + this.hitSlop);
+
+		// Reset optimal Tear slot from prior calculations
+		this.tearSlot = null;
+		
+		/*
+		 * For specs that gem ArP, determine whether the current gear
+		 * configuration will optimally hard stack Fractured gems or not.
+		 */
+		this.passiveArp = passiveStats.getStat(Stat.StatArmorPenetration);
+		this.arpStackDetected = this.detectArpStackConfiguration(ungemmedGear);
+
+		// Update red gem priority
+		const redGemCaps = new Array<GemCapsData>();
```

Comment:
The priority of adding the gems on line 328 and below can be different per class/spec. Is there a way to also make that dynamic?

Reply:
Do you have an example of a physical DPS spec that would use a different priority than the one listed here? My thought was to generalize code like this on an as-needed basis, as more sims start to implement GemOptimizer subclasses.